File: moderate/minimum_coins.cs
/* CodeEval - Minimum Coins
 * https://www.codeeval.com/browse/74/
 */

using System.IO;
using System.Collections.Generic;

class Program
{

    static int[] coins = new int[3]{5,3,1};

    static List<int> CalcNumbCoins(int amount)
    {
        List<int> coinsToReturn = new List<int>();
        foreach (int coin in coins)
        {
            float quotient = amount / (float)coin;
            if (quotient >= 1)
            {
                int numberOfCoins = (int)quotient;
                for (int i=0; i < numberOfCoins; i++)
                    coinsToReturn.Add(coin);
                amount -= numberOfCoins * coin;
            }

            if (amount == 0)
                return coinsToReturn;
        }
        return coinsToReturn;
    }

    static void Main(string[] args)
    {
        using (StreamReader reader = File.OpenText(args[0]))
        while (!reader.EndOfStream)
        {
            string line = reader.ReadLine();
            if (null == line)
                continue;

            int amount = int.Parse(line.Trim());
            List<int> numberCoins = CalcNumbCoins(amount);
            System.Console.WriteLine(numberCoins.Count);
        }
    }
}

File: moderate/minimum_coins.pl
#!/usr/bin/env perl
# CodeEval - Minimum Coins
# https://www.codeeval.com/open_challenges/74/

use strict;

my @coins = (5, 3, 1);

sub calc_numb_coins {
    my @coins_to_return = ();
    my $amount = shift @_;
    
    foreach my $coin (@coins){
        my $quotient = $amount / $coin;
        if ($quotient >= 1){
            my $number_of_coins = sprintf("%d", $quotient);
            for (my $i = 0; $i < $number_of_coins; $i++){
                #print "coin: $coin ";
                push(@coins_to_return, $coin);
            }
            $amount -= $number_of_coins * $coin;                
        }

        if (!$amount){
            return @coins_to_return;
        }
    }
}

open(INFILE, "<", $ARGV[0]) or die("Cannot open file $ARGV[0] for reading: $!");
while(my $line = <INFILE>) {
    next if($line =~ m/^\s$/);  
    $line =~ s/(^\s|\s*$)//g;   
    my $amount = $line;
    my @number_of_coins = calc_numb_coins($amount);
    print scalar @number_of_coins;
    print "\n";
}
close(INFILE);
